packages/dataviews/src/hooks/use-form-validity.ts

```diff
@@ -74,486 +76,617 @@ function isFormValid( formValidity: FormValidity | undefined ): boolean {
+					String( error ) ||
+					__( 'Unknown error when running custom validation.' );
 			}
+			return {
+				custom: {
+					type: 'invalid',
+					message: errorMessage,
+				},
+			};
+		}
+	}
 
-			// No errors for this field, remove from errors object
-			setFormValidity( ( prev ) => {
-				if ( ! prev ) {
-					return prev;
-				}
+	if ( typeof customError === 'string' ) {
+		return {
+			custom: {
+				type: 'invalid',
+				message: customError,
+			},
+		};
+	}
 
-				if ( parentFieldId ) {
-					// This field is a child - remove it from parent's children
-					const parentField = prev[ parentFieldId ];
-					if ( ! parentField?.children ) {
-						return prev;
-					}
-
-					const { [ field.id ]: removed, ...restChildren } =
-						parentField.children as any;
-
-					// If no more children, remove the children property
-					if ( Object.keys( restChildren ).length === 0 ) {
-						const { children, ...restParent } = parentField;
-						if ( Object.keys( restParent ).length === 0 ) {
-							// Remove parent field entirely if no other validations
-							const {
-								[ parentFieldId ]: removedParent,
-								...restFields
-							} = prev;
-							return Object.keys( restFields ).length === 0
-								? undefined
-								: restFields;
-						}
-						return {
-							...prev,
-							[ parentFieldId ]: restParent,
-						};
-					}
-
-					return {
-						...prev,
-						[ parentFieldId ]: {
-							...parentField,
-							children: restChildren,
-						},
-					};
-				}
+	// Validate the field: isValid.custom (async)
+	if ( customError instanceof Promise ) {
+		handleCustomValidationAsync( customError, formField, promiseHandler );
 
-				// Regular field - remove from top level
-				if ( ! prev[ field.id ] ) {
-					return prev;
-				}
+		return {
+			custom: {
+				type: 'validating',
+				message: __( 'Validating…' ),
+			},
+		};
+	}
 
-				const { [ field.id ]: removed, ...rest } = prev;
+	// Validate its children.
+	if ( formField.children ) {
+		const result: Record< string, FieldValidity | undefined > = {};
+		formField.children.forEach( ( child ) => {
+			result[ child.id ] = validateFormField( item, child, {
+				...promiseHandler,
+				path: [ ...promiseHandler.path, formField.id, 'children' ],
+			} );
+		} );
 
-				if ( Object.keys( rest ).length === 0 ) {
-					return undefined;
-				}
+		const filteredResult: Record< string, FieldValidity > = {};
+		Object.entries( result ).forEach( ( [ key, value ] ) => {
+			if ( value !== undefined ) {
+				filteredResult[ key ] = value;
+			}
+		} );
+
+		if ( Object.keys( filteredResult ).length === 0 ) {
+			return undefined;
+		}
+
+		return {
+			children: filteredResult,
+		};
+	}
+
+	// No errors for this field or its children.
+	return undefined;
+}
 
-				return rest;
+function getFormFieldValue< Item >(
+	formField: FormFieldToValidate< Item >,
+	item: Item
+): any {
+	const fieldValue = formField?.field?.getValue( { item } );
+	if ( ! formField.children ) {
```

Comment:
Nitpick: bellow we have another check formField.children.length > 0, could we combine the two checks here and simplify things a little e.g: if (!formField.children || formField.children.length === 0)